### PR TITLE
ANNULUS-325: Cache Blocks By Height

### PR DESCRIPTION
**Pull request by lukorito**
## Description

Add caching for blocks

Fixes # (issue)

## Type of change

Please delete options that are not relevant.

- [ ] New feature (non-breaking change which adds functionality)

## Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [ ] I have checked my code and corrected any misspellings
- [ ] I have added screenshots

### CODE REVIEW

File: lib/blocks/providers/block_provider.dart [hunk 1]
```diff
@@ -158,13 +164,28 @@ class BlockNotifier extends StateNotifier<AsyncValue<Map<int, Block>>> {
 
       //fetch more blocks with block0 as the reference for a fixed depth
       for (int i = 1; i < pageLimit; i++) {
-        final blockRes = await genusClient.getBlockByHeight(height: blocks[0].height - i);
-        blocks.add(
-          Block.fromBlockRes(
+        final int heightToSearch = blocks[0].height - i;
+        // check is blocks exist in cache
+        if (await HiveService().getItem(boxType: Hives.blocks, key: heightToSearch.toString()) != null) {
```
Comment (berriose) [inline lib/blocks/providers/block_provider.dart:169]: We can make this await call once and reuse the value to speed it up slightly.

File: lib/blocks/providers/block_provider.dart [hunk 2]
```diff
@@ -205,21 +226,39 @@ class BlockNotifier extends StateNotifier<AsyncValue<Map<int, Block>>> {
...
-      final blockRes = await genusClient.getBlockByHeight(height: desiredHeight);
-      final presentConfig = await config;
-      // Add that block to state's list
-      var newBlock = Block.fromBlockRes(
-        blockRes: blockRes,
-        epochLength: presentConfig.config.epochLength.toInt(),
-      );
-      blocks[depth] = newBlock;
+      // check is blocks exist in cache
+      if (await HiveService().getItem(boxType: Hives.blocks, key: desiredHeight.toString()) != null) {
```
Comment (berriose) [inline lib/blocks/providers/block_provider.dart:230]: same here

File: lib/blocks/providers/block_provider.dart [hunk 3]
```diff
@@ -292,29 +349,49 @@ class BlockNotifier extends StateNotifier<AsyncValue<Map<int, Block>>> {
...
 
-    final genusClient = ref.read(genusProvider(selectedChain));
-    var nextBlock = await genusClient.getBlockByHeight(height: height);
-    //check that block has transactions
-    while (!nextBlock.block.fullBody.hasField(1)) {
-      height--;
-      nextBlock = await genusClient.getBlockByHeight(height: height);
-    }
+    // check if item exists in cache
+    if (await HiveService().getItem(boxType: Hives.blocks, key: height.toString()) != null) {
```
Comment (berriose) [inline lib/blocks/providers/block_provider.dart:353]: same here

**Review by berriose — APPROVED** [review]

File: lib/blocks/providers/block_provider.dart [hunk 4]
```diff
@@ -292,29 +352,50 @@ class BlockNotifier extends StateNotifier<AsyncValue<Map<int, Block>>> {
...
+      final depth = blockAtDepth0.height - height;
 
-    //not in state
-    if (blocks[depth] == null) {
-      blocks = {...blocks};
-      final sortedBlocks = sortBlocksByDepth(blocks: blocks);
-      state = AsyncData(sortedBlocks);
-    }
+      //not in state
+      if (blocks[depth] == null) {
```
Comment (berriose) [inline lib/blocks/providers/block_provider.dart:381]: Did we forget to add the fetched block into state?

Comment (lukorito) [reply lib/blocks/providers/block_provider.dart:381]: It happens here https://github.com/Topl/flutter_annulus/pull/104/files/37bb1f04d192849d056094884f5085dfaf8a78ab#diff-64a63ad5b1865ec7868070386de610139dcde0999b0134094b03f26ddebeb600R384

**Pull request merged**